### ISSUE TITLE
Need to be able to tell if the caret is displayed at the beginning or the end of the line

### ISSUE THREAD

**Issue by BenjamP** [w3c/selection-api#32]
If a caret is at the end of a line that wraps, it's hard to tell where it is on the screen (end of line or beginning of the next line). This is important for being able to control the selection (beforeSelectionChange).
-From MS Office Team

**Comment by johanneswilm** [comment]
I agree. Very important for manual movement of the caret in block direction.

**Comment by rniwa** [comment]
Should we add new property like `readonly boolean attribute beforeLineWrap` on `Selection` interface?

**Comment by johanneswilm** [comment]
That should work to find out where the caret is yes. So that's a good idea.

How about this part:

> This is important for being able to control the selection (beforeSelectionChange).

As I understand it, we should also be able to set the selection at the start or end of the line, right? This would also be useful for manual movement of caret in block direction.

**Comment by rniwa** [comment]
For that, we probably need to add a new event with a sequence of `Range` objects and a boolean indicating the caret is before or after the line wrap if the range is collapsed (i.e. selection is caret), and probably add a new method on `Selection` like `setCaretPosition(Node node, unsigned long offset, bool beforeLineWrap)` that specifies whether the caret should appear before the line wrap or not if the specified DOM position happens to be at a line wrap.

**Comment by johanneswilm** [comment]
Sounds good!

**Comment by yosinch** [comment]
Do we really need to expose beforeLineWrap to web author? It makes user script complicated.
Once we can specify beforeLineWrap in Selection.modify(), e.g. move towards "inline-end" set internal state `beforeLineWrap` to true.

Once the spec specifies firing `selecitonChange` event based on caret quad change instead of caret DOM position change, this issue can be resolved.

**Comment by rniwa** [comment]
Well, the use case here is saving and restoring selection after mutating DOM so the author script must be able to tell whether `beforeLineWrap` is true or false.

**Comment by rubywwwilde** [comment]
Also, right now if you get bounding client rect of a range that is collapsed at a line break, it returns a dom rect, that is on the end of the previous line. That is completely misleading for the case when the caret stays on the start of the next line.
Therefore getBoundingClientRect should return different rects based on `beforeLineWrap`.
